kolibri/auth/management/commands/importusers.py

```diff
@@ -0,0 +1,141 @@
+            username=username,
+            facility=facility,
+        )
+        password = user.get('password', DEFAULT_PASSWORD)
+        new_user.set_password(password)
+        new_user.save()
+        if classroom:
+            classroom.add_member(new_user)
+        logging.info('User created with username {username} in facility {facility} with password {password}'.format(
+            username=username,
+            facility=facility,
+            password=password,
+        ))
+        return True
+
+class Command(BaseCommand):
+    help = """
+    Imports a user list from CSV file and creates
+    them on a specified or the default facility.
+
+    Requires CSV file data in this form:
+    <full_name>,<username>,<password>,<facility>,<class>
+
+    Less information can be passed if the headers are specified:
+    full_name,username,password,class
+    <full_name>,<username>,<password>,<class>
+
+    Or in a different order:
+    username,full_name
+    <username>,<full_name>
+
+    At minimum the username is required.
+    The password will be set to '{DEFAULT_PASSWORD}' if none is set.
+    The facility can be either the facility id or the facility name.
+    If no facility is given, either the default facility,
+    or the facility specified with the --facility commandline argument will be used.
+    """.format(DEFAULT_PASSWORD=DEFAULT_PASSWORD)
+
+    def add_arguments(self, parser):
+        parser.add_argument('filepath', action='store', help='Path to CSV file.')
+        parser.add_argument('--facility', action='store', type=str, help='Facility id to import the users into')
+
+    def handle(self, *args, **options):
+        if options['facility']:
+            default_facility = Facility.objects.get(pk=options['facility'])
+        else:
+            default_facility = Facility.get_default_facility()
+
+        if not default_facility:
+            raise CommandError('No default facility exists, please make sure to provision this device before running this command')
+
+        fieldnames = ['full_name', 'username', 'password', 'facility', 'class']
+        with open(options['filepath']) as f:
+            header = next(csv.reader(f, strict=True))
+            if all(col in fieldnames for col in header):
+                # Every item in the first row matches an item in the fieldnames, it is a header row
+                if 'username' not in header:
+                    raise CommandError('No usernames specified, this is required for user creation')
+                ordered_fieldnames = header
+            elif any(col in fieldnames for col in header):
+                raise CommandError('Mix of valid and invalid header labels found in first row')
+            else:
+                ordered_fieldnames = fieldnames
+
+        with open(options['filepath']) as f:
+            reader = csv.DictReader(f, fieldnames=ordered_fieldnames, strict=True)
+            with transaction.atomic():
```

Comment:
Nice work on the atomatic part! Was about to suggest it further up, but this is great. Maybe we could error handle this such that if it fails, we debug the name of the user that it fails to create? Am thinking that validation and database constraints may reject specific rows.

Reply:
Is the above fixed? ^

Reply:
Yes, catching validation errors in the create user function.